From c2f8fa194e7a69e046f6d5fa2735b359de1b1dd8 Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Thu, 18 Jun 2020 22:45:45 +0800
Subject: [PATCH] DRVLIB_SW-3663:GSWAPI: changing subifidgroup bridgeport id
 bit positions

subinterface id group or ingress bridgeport id single bit
changing for based on tflow table spec

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c | 7 +++++--
 include/net/switch_api/lantiq_gsw_flow.h              | 6 ++----
 2 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
index 4dc1a85a5ee0..772e7f784d52 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_pce.c
@@ -2474,14 +2474,17 @@ int pce_rule_read(void *cdev, ltq_pce_table_t *pthandle, GSW_PCE_rule_t *parm)
 				for (i = 0; i <= 7; i++)
 					parm->action.nForwardPortMap[i] = ptbl.val[i + 10];
 
-				if (((ptbl.val[4] >> 2) & 0x3) == 0x3) {
+				if ((ptbl.val[10 + gswdev->cport / 16] >> 0) == 0x1) {
+					parm->action.ePortMapAction =
+						GSW_PCE_ACTION_PORTMAP_CPU;
+				} else if (((ptbl.val[4] >> 2) & 0x3) == 0x3) {
 					//If port-map type is 'tflow'.
 					parm->action.ePortMapAction =
 						GSW_PCE_ACTION_PORTMAP_ALTERNATIVE;
 				} else if (((ptbl.val[4] >> 2) & 0x3) == 0x1) {
 					//If port-map type is 'MC router'.
 					parm->action.ePortMapAction =
-						GSW_PCE_ACTION_PORTMAP_CPU;
+						GSW_PCE_ACTION_PORTMAP_MULTICAST_ROUTER;
 				}
 			}
 		}
diff --git a/include/net/switch_api/lantiq_gsw_flow.h b/include/net/switch_api/lantiq_gsw_flow.h
index 7d037f274c40..aeffd52b03a5 100644
--- a/include/net/switch_api/lantiq_gsw_flow.h
+++ b/include/net/switch_api/lantiq_gsw_flow.h
@@ -116,12 +116,10 @@ typedef enum {
 /** \brief Select Mode of Sub-Interface ID Field.
     Used by \ref GSW_PCE_pattern_t. */
 typedef enum {
-	/** Sub Interface ID as defined by GSWIP-3.0. */
-	GSW_PCE_SUBIFID_TYPE_SUBIFID = 0,
 	/** Sub Interface ID group as defined by GSWIP-3.1. */
-	GSW_PCE_SUBIFID_TYPE_GROUP = 1,
+	GSW_PCE_SUBIFID_TYPE_GROUP = 0,
 	/** Bridge Port ID as defined by GSWIP-3.1. */
-	GSW_PCE_SUBIFID_TYPE_BRIDGEPORT = 2
+	GSW_PCE_SUBIFID_TYPE_BRIDGEPORT = 1
 } GSW_PCE_SUBIFID_TYPE_t;
 
 /** \brief Packet Classification Engine Pattern Configuration.
